### PR TITLE
[REG-1884] Don't send 2 of same key in same frame

### PR DESCRIPTION
**Pull request by RG-nAmKcAz**
- Stops us from sending 2 events for the same key within the same frame
- Adds the current frame # to event send messages for clarity/debugging
- Test: Close in time, or even intentionally making the end before the start don't write in the same frame

image

image

---

Find the pull request instructions [here](https://www.notion.so/regressiongg/Pull-Request-Process-0d57a7eb582a446983edfacafa406f1e?pvs=4)

Every reviewer and the owner of the PR should consider these points in their request (feel free to copy this checklist so you can fill it out yourself in the overall PR comment)

- [ ] The code is extensible and backward compatible
- [ ] New public interfaces are extensible and open to backward compatibility in the future
- [ ] If preparing to remove a field in the future (i.e. this PR removes an argument), the argument stays but is no longer functional, and attaches a deprecation warning. A linear task is also created to track this deletion task.
- [ ] Non-critical or potentially modifiable arguments are optional
- [ ] Breaking changes and the approach to handling them have been verified with the team (in the Linear task, design doc, or PR itself)
- [ ] The code is easy to read
- [ ] Unit tests are added for expected and edge cases
- [ ] Integration tests are added for expected and edge cases
- [ ] Functions and classes are documented
- [ ] Migrations for both up and down operations are completed
- [ ] A documentation PR is created and being reviewed for anything in this PR that requires knowledge to use
- [ ] Implications on other dependent code (i.e. sample games and sample bots) is considered, mentioned, and properly handled
- [ ] Style changes and other non-blocking changes are marked as non-blocking from reviewers

### CODE REVIEW

File: src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/BotSegments/Models/InputPlaybackActionData.cs [hunk 1]
```diff
@@ -51,10 +52,19 @@ public bool ProcessAction(int segmentNumber, Dictionary<long, ObjectStatus> curr
 
                 var allInputs = inputData.AllInputsSortedByTime();
 
+                HashSet<Key> keysSeenThisUpdate = new();
+                bool stopKeyboard = false;
+
                 foreach (object input in allInputs)
                 {
-                    if (input is KeyboardInputActionData replayKeyboardInputEntry)
+                    if (!stopKeyboard && input is KeyboardInputActionData replayKeyboardInputEntry)
```
Comment (vontell) [inline src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/BotSegments/Models/InputPlaybackActionData.cs:60]: I am potentially missing something or a little bit confused - wouldn't this stop any other key events from being sent as well? For instance, if my actions are press A, A, then B, wouldn't B fail to ever be executed?

**Review by vontell — CHANGES_REQUESTED** [review]
Question about the behavior of this, but seems sound otherwise!

File: src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/BotSegments/Models/InputPlaybackActionData.cs (hunk 1 above)
Comment (vontell) [reply src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/BotSegments/Models/InputPlaybackActionData.cs:60]: Ah I see... that makes more sense now. Thanks!

**Review by vontell — APPROVED** [review]

**Review by abeizer — APPROVED** [review]

**Review by svolokh — APPROVED** [review]
This makes sense given the way the keyboard recordings are stored. 

Perhaps it could make sense one day to explore capturing the actual underlying down/up events themselves (e.g. with [InputEventTrace](https://docs.unity3d.com/Packages/com.unity.inputsystem@1.5/manual/Events.html)), which could reduce the need for such handling.

**Pull request merged**